docs/proposals/templates.md

```diff
@@ -0,0 +1,349 @@
+  print json.dumps({
+    'apiVersion': 'v1',
+    'kind': 'Pod',
+    'metadata': {
+      'name': kubernetes_tmpl.GetVar('message'),
+      'labels': {'name': 'nginx'},
+    },
+  })
+```
+
+Where "templateBody" is arbitrary content that will be processed by the appropriate template plugin, as selected by templateKind.
+
+Because the template body is arbitrary and various plugins can be provided, the overall capability and syntax of templates 
+is not dictated by this specification.
+
+The schema also provides for an array of references to additional templates that should be expanded when this template is
+expanded.
+
+Additional example ("openshift" syntax):
+```
+kind: "TemplateDeclaration"
+apiVersion: "v1"
+metadata: 
+  name: "mongodb-ephemeral"
+  annotations: 
+    description: "Provides a MongoDB database service"
+templateKind: openshift
+parameters: 
+  - 
+    name: "DATABASE_SERVICE_NAME"
+    description: "Database service name"
+    value: "mongodb"
+    required: true
+    type: "string"
+  - 
+    name: "MONGODB_USER"
+    description: "Username for MongoDB user that will be used for accessing the database"
+    required: true
+    type: "string"
+  - 
+    name: "MONGODB_PASSWORD"
+    description: "Password for the MongoDB user"
+    required: true
+    type: "string"
+  - 
+    name: "MONGODB_DATABASE"
+    description: "Database name"
+    value: "sampledb"
+    required: true
+    type: "string"
+  - 
+    name: "REPLICA_COUNT"
+    description: "Number of mongo replicas to run"
+    value: "1"
+    required: true
+    type: "int"
+body: |
+  objects: 
+    - 
+      kind: "Service"
+      apiVersion: "v1"
+      metadata: 
+        name: "$(DATABASE_SERVICE_NAME)"
+      spec: 
+        ports: 
+          - 
+            name: "mongo"
+            protocol: "TCP"
+            targetPort: 27017
+        selector: 
+          name: "$(DATABASE_SERVICE_NAME)"
+    - 
+      kind: "ReplicationController"
+      apiVersion: "v1"
+      metadata: 
+        name: "$(DATABASE_SERVICE_NAME)"
+      spec: 
+        replicas: "$(REPLICA_COUNT)"
+        selector: 
+          name: "$(DATABASE_SERVICE_NAME)"
+        template: 
+          metadata: 
+            creationTimestamp: null
+            labels: 
+              name: "$(DATABASE_SERVICE_NAME)"
+          spec: 
+            containers: 
+              - 
+                name: "mongodb"
+                image: "docker.io/centos/mongodb-26-centos7"
+                ports: 
+                  - 
+                    containerPort: 27017
+                    protocol: "TCP"
+                env: 
+                  - 
+                    name: "MONGODB_USER"
+                    value: "${MONGODB_USER}"
+                  - 
+                    name: "MONGODB_PASSWORD"
+                    value: "${MONGODB_PASSWORD}"
+                  - 
+                    name: "MONGODB_DATABASE"
+                    value: "${MONGODB_DATABASE}"
+```
+
```

Comment:
Yeah that's two distinct choices.  The one I proposed is also the DM approach, and also closer to what Terraform does.  This choice also is linked to the choice of how nested templates are defined, since with your proposal I think you're limited to the following, but with the DM approach for invoking templates, one also has the design choice of returning invocation objects from the template body itself.

```
includes:
   - path: https://github.com/myorg/mytemplates/template1.yml
   - path: https://github.com/myorg/mytemplates/template2.yml
```


I wonder, with your proposal, in order to specify arguments, will we not end up copy/pasting templates around with nothing changed except the defaults?